scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,376 @@
+            .toIterant
+            .collect {
+              case MessageReceived(packet) => packet
+            }
+            .mapEval { packet =>
+              currentTimeMillis.flatMap { timestamp =>
+                Packet.unpack(packet) match {
+                  case Attempt.Successful((payload, remotePublicKey)) =>
+                    payload match {
+                      case _: Payload.Request =>
+                        // Not relevant on the client channel.
+                        Task.pure(None)
+
+                      case p: Payload.HasExpiration[_] if p.expiration < timestamp - expirationMillis =>
+                        Task.pure(None)
+
+                      case p: Payload.Response =>
+                        Task.pure(Some(p))
+                    }
+
+                  case Attempt.Failure(err) =>
+                    Task.raiseError(
+                      new IllegalArgumentException(s"Failed to unpack message: $err")
+                    )
+                }
+              }
+            }
+            .collect {
+              case Some(response) => response
+            }
+            .collect(pf)
+
+        /** Collect the first response that matches the partial function or return None if one cannot be found */
+        def collectFirstResponse[T](pf: PartialFunction[Payload.Response, T]): Task[Option[T]] =
+          channel
+            .collectResponses(requestTimeout.fromNow)(pf)
+            .headOptionL
+            .onErrorRecover {
+              case NonFatal(ex) => None
+            }
+
+        /** Collect responses that match the partial function and fold them while the folder function returns Left.  */
+        def collectAndFoldResponses[T, Z](timeout: FiniteDuration, seed: Z)(pf: PartialFunction[Payload.Response, T])(
+            f: (Z, T) => Either[Z, Z]
+        ): Task[Option[Z]] =
+          channel
+            .collectResponses(timeout.fromNow)(pf)
+            .attempt
+            .foldWhileLeftL((seed -> 0).some) {
+              case (Some((acc, count)), Left(ex: TimeoutException)) if count > 0 =>
+                // We have a timeout but we already accumulated some results, so return those.
+                Right(Some((acc, count)))
+
+              case (_, Left(_)) =>
+                // Unexpected error, discard results, if any.
+                Right(None)
+
+              case (Some((acc, count)), Right(response)) =>
+                // New response, fold it with the existing to decide if we need more.
+                val next = (acc: Z) => Some(acc -> (count + 1))
+                f(acc, response).bimap(next, next)
+            }
+            .map(_.map(_._1))
+
+      }
+    }
+  }
+
+  // Functions to be applied on the `.nextMessage()` or `.nextServerEvent()` results.
+  private implicit class NextOps[A](next: Task[Option[A]]) {
+    def toIterant: Iterant[Task, A] =
+      Iterant.repeatEvalF(next).takeWhile(_.isDefined).map(_.get)
```

Comment:
Iterant is entirely pull based, so there's no need for stuff like worrying about back pressure or buffering. Since we have a method that reads one message at a time, I think this is easier to reason about in this context, when we pull messages from a queue.

Reply:
It also has a killer [foldWhileLeftEvalL](https://monix.io/api/3.0/monix/tail/Iterant.html#foldWhileLeftEvalL[S](seed:F[S])(op:(S,A)=%3EF[Either[S,S]])(implicitF:cats.effect.Sync[F]):F[S]) method, although I didn't need it here.